airbyte-integrations/connectors/source-github/source_github/streams.py

```diff
@@ -0,0 +1,446 @@
+            `reviews` record. We may leave only `user.id` field and save in to `user_id` field in the record. So if you
+            need to do something similar with your record you may use this method.
+        """
+        for field in self.object_fields:
+            field_value = record.pop(field, None)
+            record[f"{field}_id"] = field_value.get("id") if field_value else None
+
+        for field in self.list_fields:
+            field_values = record.pop(field, [])
+            record[field] = [value.get("id") for value in field_values]
+
+        # TODO In the future when we will add support for multiple repositories we will need an additional field in
+        #  each stream that will tell which repository the record belongs to. In singer it was `_sdc_repository` field.
+        #  You may see an example here:
+        #  https://github.com/airbytehq/tap-github/blob/a0530b80d299864a9a106398b74d87c31ecc8a7a/tap_github/__init__.py#L253
+        #  I decided to leave this field for future usage when support for multiple repositories will be added.
+        record["_sdc_repository"] = None
+
+        return record
+
+
+class SemiIncrementalGithubStream(GithubStream):
+    cursor_field = "updated_at"
+
+    # This flag is used to indicate that current stream supports `sort` and `direction` request parameters and that
+    # we should break processing records if possible (see comment for `self._should_stop`).
+    should_break_if_sorter = False
+
+    def __init__(self, start_date: str, **kwargs):
+        super().__init__(**kwargs)
+
+        self._start_date = start_date
+
+        # This flag is used for streams which support `sort` and `direction` request parameters
+        # (`should_break_if_sorter` is set to `True`). If `sort` is set to `updated` and `direction` is set to `desc`
+        # this means that latest records will be at the beginning of the response and after we processed those latest
+        # records we can just stop and not process other record. This will increase speed of each incremental stream
+        # which supports those 2 request parameters.
+        self._should_stop = False
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]):
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state
+        object and returning an updated state object.
+        """
+        state_value = latest_cursor_value = latest_record.get(self.cursor_field)
+
+        if current_stream_state.get(self.cursor_field):
+            state_value = max(latest_cursor_value, current_stream_state[self.cursor_field])
+
+        return {self.cursor_field: state_value}
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        if not self._should_stop:
+            return super().next_page_token(response=response)
+
+    def parse_response(
```

Comment:
would it be simpler to put this logic in `read_records`? e.g: 

```
def read_records():
  for record in response: 
    if record[cursor] >= start: 
      yield record
    elif break_if_sorted: 
      return
```
this would allow us to not redefine next page token

Reply:
We can't do that because if we move this logic to `read_records` then the whole optimization will not work and we won't get anything from sorted response and we will process all records and not break where needed.

Reply:
why? Wouldn't the data returned from `super().read_records()` be sorted descending which means we can break once the current record's cursor is less than the start point cursor?

Reply:
You are right, it worked 👍

Reply:
Done.